draft-ralston-mimi-linearized-matrix.md

```diff
@@ -687,17 +687,218 @@ The hub server MUST enforce the room version's algorithms upon the room. Partici
+for which localparts (see {{int-room-id}}) are already claimed by other rooms. Having independent
+group IDs and room IDs can lead to confusion and a similar sort of namespacing issue (a room creator
+can create a conflicting group ID). Instead, the server (usually) creates the room on behalf of the
+client, allowing the client to then send the initial public group state to the room for other MLS
+members.
+
+## Updating Group State {#int-mls-add-remove}
+
+This document does not provide a way to send proposals to the MLS group, meaning all commits MUST only
+contain proposals which are sent by the same member (see {{Section 12 of I-D.ietf-mls-protocol}}).
+
+All commits are encoded as `m.mls.commit` events ({{int-ev-mls-commit}}) and are sent to the room.
+These commits are additionally encoded using `PublicMessage`, giving servers visibility on the contents
+of the commits. Upon receiving the event (see {{int-receiving-events}}), the hub server MUST additionally
+validate that any membership changes match what is possible with the room membership:
+
+* Devices can only be added to the group if they belong to a user which is joined to the room. It is
+  not enough to be invited, knocking, etc on the room - the user ID MUST be in the `join` state.
+* A device can remove another device if they both belong to the same user ID.
+* A device can be removed by anyone if the user ID to which it belongs is no longer in the `join` state.
+* A device cannot be removed any other way, as per MLS.
+
+If this validation fails, the hub server MUST reject the request if it's shaped as an LPDU ({{int-lpdu}})
+and soft-fail ({{int-soft-failure}}) the event if it's a PDU ({{int-pdu}}).
+
+Welcome messages are sent to devices over to-device messaging ({{int-transport-to-device}}). The `type`
+for the message is `m.room.encrypted` [**TODO**: Rename to avoid confusion with room event?] and `content`
+of:
+
+~~~ json
+{
+   "algorithm": "m.mls.v1.welcome.[ciphersuite]",
+   "ciphertext": "<unpadded base64 encoded welcome message>",
+   "commit_event_id": "<event ID of the m.mls.commit event>"
+}
+~~~
+
+`algorithm` is the ciphersuite, `dhkemx25519-aes128gcm-sha256-ed25519`, prefixed with `m.mls.v1.welcome`.
+
+The remaining fields are as described in the example. See {{int-unpadded-base64}} for "unpadded base64".
+
+All fields MUST be supplied. Note that the sender's user ID and device ID are made available over the
+to-device messaging endpoints ({{int-transport-to-device}}).
+
+In all cases, a device remembers the event ID (either from the `m.mls.commit` event or `commit_event_id`
+from a to-device message) after decryption to associate it with the MLS epoch. The device can then do
+a reverse lookup of epoch to event ID to MLS group state. Note that a client *always* has access to
+`m.mls.commit` events, even when hidden by history visibility ({{int-calc-event-visibility}}).
```

Comment:
Is this for supporting external joins?  I don't think it's needed for anything else.

Reply:
This is largely based on my understanding for what's needed to ensure that the membership rules above are honoured, though maybe that's something the clients can figure out on their own?

Reply:
Yeah, if the client is supposed to validate the full history, then I suppose it's needed.  I haven't really thought much about that.